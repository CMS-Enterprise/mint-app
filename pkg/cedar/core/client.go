package cedarcore

import (
	"context"
	"net/http"
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
	"github.com/cmsgov/easi-app/pkg/flags"
)

const (
	cedarCoreEnabledKey     = "cedarCoreEnabled"
	cedarCoreEnabledDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", cedarCoreEnabledKey),
				zap.Bool("flagDefault", cedarCoreEnabledDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.DefaultClient

	client := &Client{
		cedarCoreEnabled: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    hc,
		cache: c,
	}

	// Start cache refresh for systems
	client.startCacheRefresh(ctx, time.Minute*5, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	cedarCoreEnabled func(context.Context) bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARCoreAPI
	hc               *http.Client
	cache            *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime
// This function returns no errors, and only logs when something goes wrong
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				appcontext.ZLogger(ctx).Error("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
