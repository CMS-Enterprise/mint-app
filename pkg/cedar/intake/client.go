// Package intake serves as an Anti-Corruption Layer (in Domain Driven Design
// parlance), keeping the implementation details of working with the
// downstream CEDAR Intake API well separated, so the rest of the internal
// workings of the EASi code are not using anything defined by the
// external API
package intake

import (
	"context"
	"fmt"
	"net/http"
	"strconv"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	"go.uber.org/zap"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/health_check"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/gen/client/intake"
	"github.com/cmsgov/easi-app/pkg/cedar/intake/translation"
	"github.com/cmsgov/easi-app/pkg/flags"
	"github.com/cmsgov/easi-app/pkg/models"
)

const (
	emitFlagKey = "emit-to-cedar"
	emitDefault = false
)

// NewClient builds the type that holds a connection to the CEDAR Intake API
func NewClient(cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
	fnEmit := func(ctx context.Context) bool {
		// this is the conditional way of stopping us from submitting to CEDAR; see EASI-1025
		lduser := flags.Principal(ctx)
		result, err := ldClient.BoolVariation(emitFlagKey, lduser, emitDefault)
		if err != nil {
			appcontext.ZLogger(ctx).Info(
				"problem evaluating feature flag",
				zap.Error(err),
				zap.String("flagName", emitFlagKey),
				zap.Bool("flagDefault", emitDefault),
				zap.Bool("flagResult", result),
			)
		}
		return result
	}

	hc := http.DefaultClient

	return &Client{
		emitToCedar: fnEmit,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc: hc,
	}
}

// Client represents a connection to the CEDAR Intake API
type Client struct {
	emitToCedar func(context.Context) bool
	auth        runtime.ClientAuthInfoWriter
	sdk         *apiclient.CEDARIntake
	hc          *http.Client
}

// CheckConnection hits the CEDAR Intake API `/healthcheck` endpoint to verify
// that our connection is functional
func (c *Client) CheckConnection(ctx context.Context) error {
	if !c.emitToCedar(ctx) {
		return nil
	}

	params := health_check.NewHealthCheckParamsWithContext(ctx)
	params.HTTPClient = c.hc

	resp, err := c.sdk.HealthCheck.HealthCheck(params, c.auth)
	if err != nil {
		return err
	}
	if resp.Payload == nil {
		return fmt.Errorf("no response body received")
	}
	return nil
}

// PublishSystemIntake sends a system intake to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishSystemIntake(ctx context.Context, si models.SystemIntake) error {
	intakeObject := translation.TranslatableSystemIntake(si)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishBusinessCase sends a business case to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishBusinessCase(ctx context.Context, bc models.BusinessCase) error {
	intakeObject := translation.TranslatableBusinessCase(bc)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishGRTFeedback sends an item of GRT feedback to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishGRTFeedback(ctx context.Context, feedback models.GRTFeedback) error {
	intakeObject := translation.TranslatableFeedback(feedback)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishAction sends an action to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishAction(ctx context.Context, action models.Action) error {
	intakeObject := translation.TranslatableAction(action)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishNote sends a note to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishNote(ctx context.Context, note models.Note) error {
	intakeObject := translation.TranslatableNote(note)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// private method for publishing anything that satisfies the translation.IntakeObject interface to CEDAR through the Intake API
func (c *Client) publishIntakeObject(ctx context.Context, model translation.IntakeObject) error {
	// constant values for now; may become non-constant if/when we revisit handling CEDAR validation errors

	// from the Swagger (cedar_intake.json, definitions/IntakeInput/properties/version):
	// "The version associated with the object in the body. This value can be incremented in the event a transaction needs to be resubmitted."
	const objectVersion = 1

	// from the Swagger (cedar_intake.json, paths/"/intake"/post/parameters/validatePayload):
	// "Determines if schema validation of the payload is performed syncronously before persisting the record or asyncronously after the record has been persisted"
	const isValidatedSynchronously = false // false for V1 schemas; this may change when we implement V2 schemas for EASI-1614

	id := model.ObjectID()
	objectType := model.ObjectType()

	if !c.emitToCedar(ctx) {
		appcontext.ZLogger(ctx).Info("snapshot publishing disabled", zap.String("object ID", id), zap.String("object type", objectType))
		return nil
	}

	input, err := model.CreateIntakeModel()

	if err != nil {
		appcontext.ZLogger(ctx).Error(
			"problem building cedar payload",
			zap.String("object ID", id),
			zap.String("object type", objectType),
			zap.Error(err),
		)
		return nil
	}

	params := intake.NewIntakeAddParamsWithContext(ctx)
	params.HTTPClient = c.hc
	params.Body = input

	params.ValidatePayload = strconv.FormatBool(isValidatedSynchronously)

	objectVersionStr := strconv.FormatInt(objectVersion, 10)
	params.Body.Version = &objectVersionStr

	resp, err := c.sdk.Intake.IntakeAdd(params, c.auth)
	if err != nil {
		return err
	}

	// TODO: we may need to read through the results for any "duplicate entry" error codes
	// that CEDAR _may_ send us, especially if we are re-transmitting entities... We
	// don't really care if there are dupes on their side, because we do NOT want to hold
	// any state on our side about who/what/when may have transmitted a previous copy
	// of the info to CEDAR.
	// appcontext.ZLogger(ctx).Info(fmt.Sprintf("results: %v", resp.Payload))
	_ = resp

	return nil
}
