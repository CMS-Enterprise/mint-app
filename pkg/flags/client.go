package flags

import (
	"context"
	"crypto/sha256"
	"fmt"
	"time"

	"github.com/cmsgov/mint-app/pkg/appcontext"

	"gopkg.in/launchdarkly/go-sdk-common.v2/lduser"
	ld "gopkg.in/launchdarkly/go-server-sdk.v5"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfiledata"
	"gopkg.in/launchdarkly/go-server-sdk.v5/ldfilewatch"

	"github.com/cmsgov/mint-app/pkg/appconfig"
)

// DowngradeAssessmentTeamKey is the Flag Key in LaunchDarkly for downgrading assessment team users
const DowngradeAssessmentTeamKey = "downgradeAssessmentTeam"

// Config has all the parts for creating a new LD Client
type Config struct {
	Source         appconfig.FlagSourceOption
	Key            string
	Timeout        time.Duration
	FlagValuesFile string
}

// NewLaunchDarklyClient returns a client backed by Launch Darkly
func NewLaunchDarklyClient(config Config) (*ld.LDClient, error) {
	if config.Source == appconfig.FlagSourceLaunchDarkly {
		return ld.MakeClient(config.Key, config.Timeout)
	}

	defaultTimeout := 5 * time.Second

	if config.Source == appconfig.FlagSourceFile {

		ldConfig := ld.Config{
			DataSource: ldfiledata.DataSource().FilePaths(config.FlagValuesFile).Reloader(ldfilewatch.WatchFiles),
		}
		return ld.MakeCustomClient("fake_key", ldConfig, defaultTimeout)
	}

	// we default to an OFFLINE client for non-deployed environments
	return ld.MakeCustomClient("fake_offline_key", ld.Config{Offline: true}, 5*time.Second)
}

// Principal builds the LaunchDarkly user object for the
// currently authenticated principal.
func Principal(ctx context.Context) lduser.User {
	p := appcontext.Principal(ctx)
	key := UserKeyForID(p.ID())

	// this is a bit of a loose inference, assuming a user w/o Job Codes
	// is an Anonymous user. Over time, may want to consider adding
	// a `func Anonymous() bool` accessor to the authorizaion.Principal interface
	// definition instead of doing this inference
	authed := (p.AllowUSER() || p.AllowASSESSMENT())

	return lduser.
		NewUserBuilder(key).
		Anonymous(!authed).
		Build()
}

// UserKeyForID generates a user key from an ID
// we should not be using bare EUA IDs as identifiers to
// LaunchDarkly (per Jimil/ISSO), so we use a cryptographically
// secure one-way hash of the EUA ID as "key" for the LD User object.
func UserKeyForID(id string) string {
	h := sha256.New()
	_, _ = h.Write([]byte(id))
	return fmt.Sprintf("%x", h.Sum(nil))
}
