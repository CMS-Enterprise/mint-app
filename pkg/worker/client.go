package worker

import (
	"fmt"

	faktory "github.com/contribsys/faktory/client"
)

// NewClient returns a new faktory Client.
func NewClient() (*faktory.Client, error) {
	cl, err := faktory.Open()
	if err != nil {
		return nil, fmt.Errorf("error opening faktory client: %w", err)
	}
	return cl, nil
}
