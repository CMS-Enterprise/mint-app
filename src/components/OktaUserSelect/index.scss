.usa-combo-box__list-option {
  label {
    margin-top: 0;
  }
}
