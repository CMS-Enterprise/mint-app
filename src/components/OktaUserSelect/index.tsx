import React, { CSSProperties, useEffect, useRef, useState } from 'react';
import { useTranslation } from 'react-i18next';
import Select, {
  ClearIndicatorProps,
  components,
  IndicatorsContainerProps,
  InputProps,
  MenuProps,
  OptionProps
} from 'react-select';
import { Icon } from '@trussworks/react-uswds';
import classNames from 'classnames';

import Spinner from 'components/Spinner';
import useDebounce from 'hooks/useDebounce';
import useOktaUserLookup, { OktaUserType } from 'hooks/useOktaUserLookup';
import color from 'utils/uswdsColor';

import './index.scss';

type OktaUserSelectProps = {
  className?: string;
  id: string;
  name: string;
  ariaLabelledBy?: string;
  ariaDescribedBy?: string;
  value?: OktaUserType | null;
  onChange: (contact: OktaUserType | null) => void;
  disabled?: boolean;
  autoSearch?: boolean;
};

type OktaUserSelectOption = {
  value: OktaUserType;
  label: string;
};

// Override React Select input to fix hidden input on select bug
const Input = (props: InputProps<OktaUserSelectOption, false>) => {
  return (
    <components.Input
      {...props}
      isHidden={false}
      data-testid="cedar-contact-select"
    />
  );
};

// Custom option component
const Option = (props: OptionProps<OktaUserSelectOption, false>) => {
  const { isFocused } = props;
  return (
    <components.Option
      {...props}
      className={classNames('usa-combo-box__list-option', {
        'usa-combo-box__list-option--focused': isFocused
      })}
    />
  );
};

const NoOptionsMessage = (props: any) => {
  const { t } = useTranslation('general');
  return (
    <components.NoOptionsMessage {...props}>
      <button
        type="button"
        id="no-results"
        tabIndex={0}
        className="usa-button--unstyled text-no-underline text-black"
        aria-label={t('noResults')}
      >
        {t('noResults')}
      </button>
    </components.NoOptionsMessage>
  );
};

const Menu = (props: MenuProps<OktaUserSelectOption, false>) => {
  const {
    selectProps: { inputValue }
  } = props;
  if (inputValue.length < 2) return null;
  return <components.Menu {...props} />;
};

const ClearIndicator = (
  props: ClearIndicatorProps<OktaUserSelectOption, false>
) => {
  const {
    selectProps: { inputValue, id },
    clearValue
  } = props;
  // Fix bug in 'same as requester' checkboxes where clear indicator shows with no input value
  if (!inputValue) return null;
  return (
    <button
      type="button"
      id="clear-selection"
      tabIndex={0}
      onClick={() => {
        clearValue();
        document?.getElementById(`react-select-${id}-input`)?.focus();
      }}
      className="usa-button--unstyled"
      aria-label="Clear selection"
    >
      <components.ClearIndicator {...props} />
    </button>
  );
};

const IndicatorsContainer = (
  props: IndicatorsContainerProps<OktaUserSelectOption, false>
) => {
  const {
    children,
    selectProps: { className, isDisabled },
    hasValue
  } = props;

  // Whether to show spinner based on className
  const loading = className!
    .split(' ')
    .includes('cedar-contact-select__loading');
  // Whether to show warning icon based in className
  const resultsWarning =
    hasValue && className!.split(' ').includes('cedar-contact-select__warning');

  // Hide indicators if field is disabled
  if (isDisabled) return null;

  return (
    <components.IndicatorsContainer {...props}>
      {!loading && resultsWarning && (
        <Icon.Warning className="text-warning" size={3} />
      )}
      {loading && <Spinner size="small" className="margin-right-1" />}
      {children}
    </components.IndicatorsContainer>
  );
};

/** Returns formatted contact label */
const formatLabel = (contact: OktaUserType) =>
  `${contact.displayName}${contact?.username && `, ${contact.username}`}`;

/**
 * Combobox to look up contact by name from Okta
 */
export default function OktaUserSelect({
  className,
  id,
  name,
  ariaLabelledBy,
  ariaDescribedBy,
  value,
  onChange,
  disabled,
  autoSearch
}: OktaUserSelectProps) {
  // If autoSearch, set name as initial search term
  const [searchTerm, setSearchTerm] = useState<string | undefined>(
    value ? formatLabel(value) : undefined
  );

  const [userSelected, setUserSelected] = useState(false);

  const { debounceValue, debounceLoading } = useDebounce(
    searchTerm,
    250,
    userSelected
  );

  const { contacts, queryOktaUsers, loading } = useOktaUserLookup(
    debounceValue,
    userSelected
  );

  // Selected contact
  const selectedContact = useRef(value?.username);

  /** Update contact and reset search term */
  const updateContact = (contact?: OktaUserType | null) => {
    onChange(contact || null);
    selectedContact.current = contact?.username;
    setSearchTerm(contact ? formatLabel(contact) : '');
  };

  useEffect(() => {
    if (debounceValue) {
      queryOktaUsers(debounceValue.split(',')[0]);
    }
  }, [debounceValue, queryOktaUsers]);

  // React Select styles object
  const customStyles: {
    [index: string]: (
      provided: CSSProperties,
      state: { isFocused: boolean }
    ) => CSSProperties;
  } = {
    control: (provided, state) => ({
      ...provided,
      minHeight: '40px',
      borderColor: color('base-dark'),
      outline: state.isFocused ? `.25rem solid ${color('blue-vivid-40')}` : '',
      borderRadius: 0,
      transition: 'none',
      '&:hover': {
        borderColor: color('base-dark'),
        cursor: 'text'
      }
    }),
    dropdownIndicator: provided => ({
      ...provided,
      color: color('base-dark'),
      padding: '8px 6px',
      '&:hover': {
        color: color('base-dark'),
        cursor: 'pointer'
      },
      '> svg': {
        width: '22px',
        height: '22px'
      }
    }),
    clearIndicator: provided => ({
      ...provided,
      color: color('base-dark'),
      padding: '8px 6px',
      '&:hover': {
        color: color('base-dark'),
        cursor: 'pointer'
      },
      '> svg': {
        width: '22px',
        height: '22px'
      }
    }),
    indicatorSeparator: provided => ({
      ...provided,
      marginTop: '8px',
      marginBottom: '8px'
    }),
    menu: provided => ({
      ...provided,
      marginTop: '0px',
      borderRadius: 0,
      border: `1px solid ${color('base-dark')}`,
      borderTop: 'none',
      boxShadow: 'none'
    }),
    menuList: provided => ({
      ...provided,
      paddingTop: 0,
      paddingBottom: 0
    }),
    noOptionsMessage: provided => ({
      ...provided,
      textAlign: 'left'
    }),
    input: provided => ({
      ...provided,
      visibility: 'visible'
    }),
    option: provided => ({
      ...provided,
      backgroundColor: 'transparent',
      color: 'inherit'
    })
  };

  return (
    <Select
      id={id}
      name={name}
      className={classNames(
        'margin-top-1',
        'cedar-contact-select',
        'maxw-none',
        'usa-combo-box',
        {
          'cedar-contact-select__loading':
            (loading || debounceLoading) &&
            ((searchTerm && searchTerm.length > 0) || loading) &&
            !userSelected
        },
        className
      )}
      isDisabled={disabled}
      aria-describedby={ariaDescribedBy}
      aria-disabled={disabled}
      aria-labelledby={ariaLabelledBy}
      // Conditional on menu open to override "No results" selection being tabbable/not closed on focus lost
      // Menu is closed on user selection
      menuIsOpen={
        ((contacts.length === 0 && (searchTerm?.length || 0) > 2) ||
          !!contacts.length) &&
        !userSelected
      }
      components={{
        Input,
        IndicatorsContainer,
        ClearIndicator,
        Option,
        Menu,
        NoOptionsMessage: disabled ? () => null : NoOptionsMessage
      }}
      options={contacts.map((contact: OktaUserType) => ({
        label: `${contact.displayName}, ${contact.username}`,
        value: contact
      }))}
      styles={customStyles}
      defaultValue={
        value
          ? {
              value,
              label: `${value?.displayName}${
                value?.username && `, ${value?.username}`
              }`
            }
          : undefined
      }
      value={value ? { value, label: formatLabel(value) } : undefined}
      onChange={(item: OktaUserSelectOption) =>
        updateContact(item?.value || null)
      }
      onInputChange={(
        newValue: string | undefined,
        { action }: { action: string }
      ) => {
        if (action !== 'input-blur' && action !== 'menu-close') {
          // If user selected a value, no need to query and debounce again
          if (action === 'set-value') {
            setUserSelected(true);
          } else {
            setUserSelected(false);
            setSearchTerm(newValue);
          }
        }
      }}
      defaultInputValue={searchTerm}
      inputValue={searchTerm}
      classNamePrefix="cedar-contact-select"
      instanceId={id}
      placeholder={false}
      backspaceRemovesValue={false}
      controlShouldRenderValue={false}
      isSearchable
      isClearable
    />
  );
}
