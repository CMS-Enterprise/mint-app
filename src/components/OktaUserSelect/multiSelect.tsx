import React, { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import Select from 'react-select';
import classNames from 'classnames';

import Alert from 'components/shared/Alert';
import {
  ClearIndicator,
  customStyles,
  MultiSelectTag,
  Option
} from 'components/shared/MultiSelect';
import useDebounce from 'hooks/useDebounce';
import useOktaUserLookup, { OktaUserType } from 'hooks/useOktaUserLookup';

import './index.scss';

type MultiSelectOptionProps = {
  value: string;
  label: string;
  email: string;
};

const selectedContainsNonCMS = (
  selectedUsers: MultiSelectOptionProps[]
): boolean => {
  return !!selectedUsers.find(user => !user.email?.includes('cms.hhs.gov'));
};

/**
 * MINT OktaMultiselect.
 * Uses `react-select/Select` and `@trussworks/react-uswds/Tag`.
 *
 * https://www.figma.com/file/5y4EbRmFUB7xRBKUG4qlup/USWDS-Library?node-id=869%3A7346&t=WrUjXtNxIxMgpPss-0
 */
const OktaMultiSelect = ({
  id,
  inputId,
  name,
  selectedLabel,
  onChange,
  className,
  ariaLabel
}: {
  id?: string;
  inputId?: string;
  name: string;
  selectedLabel?: string;
  onChange: (values: string[]) => void;
  className?: string;
  ariaLabel: string;
}) => {
  const { t: generalReadOnlyT } = useTranslation('generalReadOnly');

  const [selected, setSelected] = useState<MultiSelectOptionProps[]>([]);

  // If autoSearch, set name as initial search term
  const [searchTerm, setSearchTerm] = useState<string | undefined>(undefined);

  const [userWarning, setUserWarning] = useState<boolean>(false);

  const [userSelected, setUserSelected] = useState<boolean>(false);

  const { debounceValue, debounceLoading } = useDebounce(
    searchTerm,
    250,
    userSelected
  );

  const { contacts, queryOktaUsers, loading } = useOktaUserLookup(
    debounceValue,
    userSelected
  );

  useEffect(() => {
    if (debounceValue) {
      queryOktaUsers(debounceValue.split(',')[0]);
    }
  }, [debounceValue, queryOktaUsers]);

  useEffect(() => {
    setUserWarning(selectedContainsNonCMS(selected));
  }, [selected]);

  return (
    <div>
      <Select
        id={id}
        inputId={inputId}
        name={name}
        className={classNames(
          'easi-multiselect usa-combo-box margin-top-1',
          {
            'cedar-contact-select__loading':
              (loading || debounceLoading) &&
              ((searchTerm && searchTerm.length > 0) || loading) &&
              !userSelected
          },
          className
        )}
        isClearable
        menuIsOpen={
          (contacts.length === 0 && (searchTerm?.length || 0) > 2) ||
          (!!contacts.length && !userSelected)
        }
        options={contacts.map(
          (contact: OktaUserType): MultiSelectOptionProps => ({
            label: `${contact.displayName}, ${contact.username} (${contact.email})`,
            value: contact.username,
            email: contact.email
          })
        )}
        onInputChange={(newValue, { action }) => {
          if (action !== 'input-blur' && action !== 'menu-close') {
            // If user selected a value, no need to query and debounce again
            if (action !== 'set-value') {
              setUserSelected(false);
              setSearchTerm(newValue);
            }
          } else {
            setUserSelected(true);
            setSearchTerm(undefined);
          }
        }}
        components={{ ClearIndicator, Option }}
        isMulti
        hideSelectedOptions={false}
        closeMenuOnSelect={false}
        tabSelectsValue={false}
        onChange={selectedOptions => {
          // TODO: Figure out what's happening with typescript here
          // @ts-ignore
          setSelected(selectedOptions);
          // @ts-ignore
          onChange(selectedOptions.map(option => option.value));
        }}
        value={selected}
        controlShouldRenderValue={false}
        placeholder={`${selected.length} selected`}
        styles={customStyles}
        aria-labelledby={ariaLabel}
        classNamePrefix="cedar-contact-select"
        instanceId={id}
        backspaceRemovesValue={false}
        isSearchable
      />
      {selected.length > 0 && (
        <div className="easi-multiselect--selected">
          <h4 className="text-normal margin-bottom-1">
            {selectedLabel || 'Selected options'}
          </h4>
          <ul className="usa-list--unstyled" id={`${id}-tags`}>
            {selected.map(({ value, label }) => (
              <li
                className="margin-bottom-05 margin-right-05 display-inline-block"
                key={value}
              >
                <MultiSelectTag
                  id={`selected-${value}`}
                  parentId={`${id}-tags`}
                  key={value}
                  label={label}
                  handleRemove={() => {
                    const updatedValues = selected.filter(
                      option => option.value !== value
                    );
                    setSelected(updatedValues);
                    onChange(updatedValues.map(option => option.value));
                  }}
                />
              </li>
            ))}
          </ul>
        </div>
      )}
      {userWarning && (
        <Alert type="warning" slim>
          {generalReadOnlyT('modal.shareAlert')}
        </Alert>
      )}
    </div>
  );
};

export default OktaMultiSelect;
