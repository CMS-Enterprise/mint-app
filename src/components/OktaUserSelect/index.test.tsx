import React from 'react';
import { MockedProvider } from '@apollo/client/testing';
import { SearchOktaUsersDocument } from 'gql/generated/graphql';
import setup from 'tests/setup';

import OktaUserSelect from './index';

describe('OktaUserSelect', () => {
  // Cedar contacts query mock
  const oktaUsersQuery = {
    request: {
      query: SearchOktaUsersDocument,
      variables: {
        searchTerm: 'Adeline'
      }
    },
    result: {
      data: {
        searchOktaUsers: [
          {
            displayName: 'Adeline Aarons',
            username: 'ABCD',
            email: 'adeline.aarons@local.fake'
          }
        ]
      }
    }
  };

  it('selects contact from dropdown', async () => {
    const { user, asFragment, getByTestId, findByText } = setup(
      <MockedProvider mocks={[oktaUsersQuery]} addTypename={false}>
        <OktaUserSelect
          id="cedarContactSelect"
          name="cedarContactSelect"
          onChange={() => null}
        />
      </MockedProvider>
    );

    // Type first name into select field input
    const input = getByTestId('cedar-contact-select');
    await user.type(input, 'Adeline');

    // Get mocked Okta result
    const userOption = await findByText('Adeline Aarons, ABCD');
    expect(userOption).toBeInTheDocument();

    // Check that component matches snapshot with expanded dropdown
    expect(asFragment()).toMatchSnapshot();

    // Select option
    await user.click(userOption);

    // Check that select field displays correct value
    expect(input).toHaveValue('Adeline Aarons, ABCD');
  });
});
