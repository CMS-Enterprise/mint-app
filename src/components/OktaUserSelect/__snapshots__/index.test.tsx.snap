// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OktaUserSelect > selects contact from dropdown 1`] = `
<DocumentFragment>
  <div
    class="margin-top-1 cedar-contact-select maxw-none usa-combo-box css-b62m3t-container"
    id="cedarContactSelect"
  >
    <span
      class="css-1f43avz-a11yText-A11yText"
      id="react-select-cedarContactSelect-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
      role="log"
    >
      <span
        id="aria-selection"
      />
      <span
        id="aria-focused"
      />
      <span
        id="aria-results"
      >
        1 result available for search term Adeline.
      </span>
      <span
        id="aria-guidance"
      >
        Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <div
      class="cedar-contact-select__control cedar-contact-select__control--is-focused cedar-contact-select__control--menu-is-open css-13u0srz-control"
    >
      <div
        class="cedar-contact-select__value-container css-1fdsijx-ValueContainer"
      >
        <div
          class="cedar-contact-select__input-container css-166bipr-Input"
          data-value="Adeline"
        >
          <input
            aria-activedescendant="react-select-cedarContactSelect-option-0"
            aria-autocomplete="list"
            aria-controls="react-select-cedarContactSelect-listbox"
            aria-describedby="react-select-cedarContactSelect-placeholder"
            aria-expanded="true"
            aria-haspopup="true"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="cedar-contact-select__input"
            data-testid="cedar-contact-select"
            id="react-select-cedarContactSelect-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value="Adeline"
          />
        </div>
      </div>
      <div
        class="cedar-contact-select__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <span
          class="cedar-contact-select__indicator-separator css-1u9des2-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="cedar-contact-select__indicator cedar-contact-select__dropdown-indicator css-ylv2i7-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="cedar-contact-select__menu css-1v1oxpf-menu"
    >
      <div
        aria-multiselectable="false"
        class="cedar-contact-select__menu-list css-bgzggo-MenuList"
        id="react-select-cedarContactSelect-listbox"
        role="listbox"
      >
        <div
          aria-disabled="false"
          aria-selected="false"
          class="usa-combo-box__list-option usa-combo-box__list-option--focused cedar-contact-select__option cedar-contact-select__option--is-focused css-10wo9uf-option"
          id="react-select-cedarContactSelect-option-0"
          role="option"
          tabindex="-1"
        >
          Adeline Aarons, ABCD
        </div>
      </div>
    </div>
    <input
      name="cedarContactSelect"
      type="hidden"
      value=""
    />
  </div>
</DocumentFragment>
`;
